ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestinationSync.java

```diff
@@ -295,10 +300,88 @@ public CompletableFuture<Void> downloadBlobAsync(BlobId blobId, OutputStream out
+        Duration.ofMillis(cloudConfig.cloudRequestTimeout), Context.NONE);
+  }
+
+  /**
+   * Returns blob properties from Azure
+   * @param blobLayout BlobLayout
+   * @return Blob properties
+   * @throws CloudStorageException
+   */
+  protected BlobProperties getBlobProperties(AzureBlobLayoutStrategy.BlobLayout blobLayout)
+      throws CloudStorageException {
+    try {
+      return createOrGetBlobStore(blobLayout.containerName).getBlobClient(blobLayout.blobFilePath).getProperties();
+    } catch (Throwable t) {
+      // Due to legacy CloudBlobStore, we have to handle 404 as a special case when it can actually be a general case.
+      if (t instanceof BlobStorageException
+          && ((BlobStorageException) t).getErrorCode() == BlobErrorCode.BLOB_NOT_FOUND) {
+        String error = String.format("Failed to get blob properties for %s from Azure blob storage because it does not exist", blobLayout);
+        logger.error(error);
+        throw new CloudStorageException(error, new StoreException(error, StoreErrorCodes.ID_Not_Found), CloudBlobStore.STATUS_NOT_FOUND, false, null);
+      }
+      String error = String.format("Failed to get blob properties for %s from Azure blob storage because %s", blobLayout, t.getMessage());
+      logger.error(error);
+      throw new CloudStorageException(error, new StoreException(error, StoreErrorCodes.Unknown_Error));
+    }
+  }
+
   @Override
   public boolean deleteBlob(BlobId blobId, long deletionTime, short lifeVersion,
-      CloudUpdateValidator cloudUpdateValidator) throws CloudStorageException {
-    return false;
+      CloudUpdateValidator unused) throws CloudStorageException {
+    Timer.Context storageTimer = azureMetrics.blobUpdateDeleteTimeLatency.time();
+    AzureBlobLayoutStrategy.BlobLayout blobLayout = azureBlobLayoutStrategy.getDataBlobLayout(blobId);
+    String blobIdStr = blobLayout.blobFilePath;
+    Map<String, String> newMetadata = new HashMap<>();
+    newMetadata.put(CloudBlobMetadata.FIELD_DELETION_TIME, String.valueOf(deletionTime));
+    newMetadata.put(CloudBlobMetadata.FIELD_LIFE_VERSION, String.valueOf(lifeVersion));
+    BlobProperties blobProperties = getBlobProperties(blobLayout);
+    Map<String, String> cloudMetadata = blobProperties.getMetadata();
+    // lifeVersion must always be present
+    short cloudlifeVersion = Short.parseShort(cloudMetadata.get(CloudBlobMetadata.FIELD_LIFE_VERSION));
+    // This is the correct behavior. For ref, look at BlobStore::delete and ReplicaThread::applyUpdatesToBlobInLocalStore
+    if (cloudlifeVersion == lifeVersion && cloudMetadata.containsKey(CloudBlobMetadata.FIELD_DELETION_TIME)
+        && Long.parseLong(cloudMetadata.get(CloudBlobMetadata.FIELD_DELETION_TIME)) != Utils.Infinite_Time ) {
+      String error = String.format("Cannot delete id %s since it is already marked as deleted in cloud", blobIdStr);
+      throw new CloudStorageException(error, new StoreException(error, StoreErrorCodes.ID_Deleted));
```

Comment:
applyDelete L1456 handles it

Reply:
oh okay. We also seem to return false (instead of throwing exception) if validation fails against the cloud metadata [here](https://github.com/linkedin/ambry/blob/d12d8265376436f112be67f0de8336e7673e85b4/ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureBlobDataAccessor.java#L349). But, we should be good as long as we catch this exception in replication and let the replication proceed. Just wanted to note.

Reply:
Used the cloud validator and threw an exc for a couple of reasons. Lmk if u disagree.

I found the behaviour of V1 to be inconsistent. 
If recentBlobCache is enabled and it has the required entry, then deleteIfNeeded throws an exc at L717.
If recentBlobCache is disabled OR it does not have the required entry, then deleteIfNeeded returns false at L709.
This is confusing esp from a testing perspective and in prod, cache evictions can happen any time.
V1 doesn't print a proper log explaining why it skipped a delete-msg. 
It just prints this `ERROR Delete failed partition 0 statusCode 500 cause StoreException made 1 attempts.`

To ease testing and to maintain a consistent behavior across all Stores, I am throwing an exc.
It will be absorbed by applyDelete L1456. 
And for testing, we catch the exc and know for sure the right code was executed.